### PR TITLE
feature NewArmyPage completed

### PR DESCRIPTION
**Pull request by cawensley**
Added 4 new pages (view armies, view army, new army, edit army) and redux variables to keep track of everything.

### CODE REVIEW

File: src/pages/Armies/EditArmyPage.test.jsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+const mockRoles = [{ id: '1111', Name: 'HQ', SortOrder: 0 }];
+const mockUser = { uid: '2222' };
+const mockCodex = 'zzzzzzz';
+const mockSquad = [{
+  id: 'TigerSquad', Codex: 'zzzzzzz', Name: 'PirateSquad11', Role: '1111', MinSize: 1, MaxSize: 5, Units: ['TigerLeader'],
+}];
+const mockUnit = [{ id: 'TigerLeader', Name: 'PirateUnit11', Gear: ['33333', '44444'] }];
+const mockEquipment = [{ id: '33333', Name: 'mockSword' }, { id: '44444', Name: 'mockGun' }];
```
Comment (ashblue) [inline src/pages/Armies/EditArmyPage.test.jsx]: All of these mocks feel really cluttered here. Probably should be in the before each block.

File: src/pages/Armies/NewArmyPage.test.jsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  const submitButton = findByTestAttr(wrapper, 'submitButton');
+  submitButton.simulate('submit', { preventDefault() {} });
+  expect(firestore().collection).toHaveBeenCalledWith('armies');
+});
+test('it should render', () => {
+  store.dispatch(RolesChange(mockRoles));
+  store.dispatch(CodexChange(mockCodex));
+  store.dispatch(SquadsChange(mockSquad));
+  store.dispatch(UnitsChange(mockUnit));
+  store.dispatch(EquipmentChange(mockEquipment));
```
Comment (ashblue) [inline src/pages/Armies/NewArmyPage.test.jsx]: Are all of these needed to demonstrate the minimal test case? Should always write the minimal amount of code to make a test fail and write the test before you write the code. It might be required but just looks odd to me on the outside.

File: src/pages/Armies/ViewArmiesPage.jsx [hunk 3]
```diff
@@ -1,14 +1,44 @@
 import React from 'react';
+import { Link } from 'react-router-dom';
 import PageTitle from '../../atoms/PageTitle';
-import CodexFilter from '../../molecules/CodexFilter';
 import RedirectButton from '../../atoms/RedirectButton';
+import DeleteButton from '../../molecules/DeleteButton';
+import store from '../../Redux/store';
 
 function ViewArmiesPage() {
```
Comment (ashblue) [inline src/pages/Armies/ViewArmiesPage.jsx]: Shouldn't this be written in ES6 without the return?

```
const ViewArmiesPage = () => (
```

File: src/pages/Armies/ViewArmyPage.jsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      <div className="row justify-content-center mt-5">
+        <div className="col-xl-8">
+          <div className="row h4 text-warning">
+            <div className="col-2">Squad</div>
+            <div className="col-2">Unit Qty</div>
+            <div className="col-2">Unit</div>
+            <div className="col-3">Unit Abilities</div>
+            <div className="col-3">Gear</div>
+          </div>
+          {displayArmy.SquadArray.map((roleArray) => (
```
Comment (ashblue) [inline src/pages/Armies/ViewArmyPage.jsx]: The code inside the map block should probably be a component

File: src/utilities/findPreviousSquad.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+import store from '../Redux/store';
+
+function findPreviousSquad(roleIndex, rowIndex) {
```
Comment (ashblue) [inline src/utilities/findPreviousSquad.js]: Should probably bundle these utilities into less files. Seems like a lot of utility files when there is common overlap to bundle them under a file name.

File: src/utilities/handleArmySubmission.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import firebase from 'firebase/app';
+import store from '../Redux/store';
+import 'firebase/firestore';
+
+export default function handleArmySubmission(input) {
```
Comment (ashblue) [inline src/utilities/handleArmySubmission.js]: I'd personally break up the logic in the if methods into private non exported methods.

File: .eslintrc.js [hunk 7]
```diff
@@ -23,5 +23,6 @@ module.exports = {
     'react',
   ],
   rules: {
+    "react/no-array-index-key": 0
```
[the remark this answers is not in the document]
Comment (cawensley) [reply .eslintrc.js]: I have several simple map functions, for example where I have a section of 'Troops' and if they user wants 2 Necron warrior squads, then I can't use unique ID's as key's because they have the same key.  So I used unique ID's as keys when I could, but instances like this where I don't have or can't use unique ID's in the map function, that is why I had that rule ignored.  What do you suggest?

File: src/Redux/actions/ArmySquadChoice.test.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+const mockPayload = { roleIndex: 0, rowIndex: 0, SquadID: 'PirateSquad' };
+const mockRole = [{ id: '99999', Name: 'BestRole' }];
+
+test('Redux Action ArmySquadChoice is successful in redux state change', () => {
+  store.dispatch(RolesChange(mockRole));
+  handleInitialArmy();
+  store.dispatch(ArmySquadRowAdd(mockRowAdd));
+  store.dispatch(ArmySquadChoice(mockPayload));
+  const ArmyState = store.getState().army.SquadArray;
+  expect(ArmyState[0].Squads[0].Squad).toBe('PirateSquad');
```
[the remark this answers is not in the document]
Comment (cawensley) [reply src/Redux/actions/ArmySquadChoice.test.js]: For all my tests on the website you would like a space before the expect?  I have 190 tests

File: src/Redux/reducers/army.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      Squad: action.payload.firstSquadID,
+      UnitQTY: action.payload.firstUnitQTY,
+      Unit: action.payload.firstUnitID,
+      Equipment: action.payload.firstEquipID,
+    };
+    const newSquads = state.SquadArray;
+    newSquads[action.payload.roleIndex].Squads.push(squadRow);
+    return { ...state, SquadArray: newSquads };
+  }
+  if (action.type === 'Army_Squad_Row_Rem') {
```
[the remark this answers is not in the document]
Comment (cawensley) [reply src/Redux/reducers/army.js]: 'Rem' was short for 'remove', but yeah i'll change it to 'remove' to make it easier to understand

File: src/atoms/PointsRow.jsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+          const gearCost = store.getState().equipment
+            .filter((data) => data.id.includes(squad.Equipment))[0].Cost;
+          totalPoints += unitQTY * gearCost;
+        }
+      });
+    });
+    store.dispatch(ArmyPointsChange(totalPoints));
+  }
+
+  // eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (cawensley) [reply src/atoms/PointsRow.jsx]: what is suspicious?

File: src/organisms/EquipmentChoice.jsx [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import PropTypes from 'prop-types';
+import store from '../Redux/store';
+import ArmyEquipmentChoice from '../Redux/actions/ArmyEquipmentChoice';
+import findPreviousUnit from '../utilities/findPreviousUnit';
+import findRealEquipmentName from '../utilities/findRealEquipmentName';
+
+function EquipmentChoice({ roleIndex, rowIndex }) {
+  const choices = findPreviousUnit(roleIndex, rowIndex)[0].Gear;
+
+  // eslint-disable-next-line
```
[the remark this answers is not in the document]
Comment (cawensley) [reply src/organisms/EquipmentChoice.jsx]: the eslint disable line?

File: src/pages/Armies/ViewArmiesPage.jsx (hunk 3 above)
Comment (cawensley) [reply src/pages/Armies/ViewArmiesPage.jsx]: For this website I made all my pages as functions that returned something.  Is that not correct?  For my last movie website I made all the pages as:  const Favoritespage = () => {......, return X }, so they still had returns in them.  Example:   https://github.com/cawensley/moviegallery/blob/master/src/pages/favorites.jsx

File: src/pages/Armies/ViewArmyPage.jsx (hunk 4 above)
Comment (cawensley) [reply src/pages/Armies/ViewArmyPage.jsx]: All the of the display pages (army, squads, armies, units, equipment) have custom displays to make the data look as best as I could.  The reason I didn't make it a component was because it's only used once.  What advantage would there be from making it a component?

File: src/pages/Armies/NewArmyPage.test.jsx (hunk 2 above)
Comment (cawensley) [reply src/pages/Armies/NewArmyPage.test.jsx]: This test is a little confusing, because you are correct that the minimum code needed to 'render' is small, but I made sure to include all those other redux dispatch methods so that all the lines of code would be run and show 100% for the testing coverage.  Is there another way I should have done that?  Split the tests into 2, one for minimum render, and a second for 100% test coverage?

File: .eslintrc.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (ashblue) [reply .eslintrc.js]: Hmmm, what about making each squad a unique ID from the database? There is also this great little utility I use for generating unique IDs locally https://www.npmjs.com/package/uuid

File: src/Redux/actions/ArmySquadChoice.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ashblue) [reply src/Redux/actions/ArmySquadChoice.test.js]: Well moving forward you should try to space your tests out based on the three As of testing for readabilty (Arrange, Act, Assert). For example.

```javascript
// Variables up here

// Interact with varaibles

// Assertion goes here
```

Spaces are super important for readability between code chunks. Line breaks give people unfamiliar with something a quick way to process it in chunks. For example, a book without paragraphs would be frustrating to read. I guess the point here is you should move forward with trying to use line breaks to separate out your code visually. I really struggled to quickly read some of the tests because of it.

File: src/pages/Armies/NewArmyPage.test.jsx (hunk 2 above)
Comment (ashblue) [reply src/pages/Armies/NewArmyPage.test.jsx]: Hmmm, well if you write tests first you'd hypothetically get 100% coverage. But this is sometimes where getting 100% coverage can cause devs to write more code than necessary. For open source projects I run 100%, but for group projects I usually do 85%. Not sure if that's an answer to the question though 🤔 You can always break up tests into smaller tests like you said to prevent things from getting hard to follow. Remember smaller tests that cover lots of code is always the goal. Because maintaining complex tests is always a nightmare.

File: src/pages/Armies/ViewArmiesPage.jsx (hunk 3 above)
Comment (ashblue) [reply src/pages/Armies/ViewArmiesPage.jsx]: Well, I meant writing functions the more modern way. Avoiding `function` wherever possible is usually a good idea.

File: src/pages/Armies/ViewArmyPage.jsx (hunk 4 above)
Comment (ashblue) [reply src/pages/Armies/ViewArmyPage.jsx]: Well smaller files make things more readable and easy to follow. Pretty much whenever I start looping over something complex I always try to move it to smaller nested components. Just easier to maintain and read long term. This file is fine for the size, but larger files (300) shoudl absolutely start breaking things off into smaller components. Although when you hit a 300 line 300 it's something too late and things have to be rewritten to fix the problem.
